local/modules/TheliaSmarty/Template/Plugins/UrlGenerator.php

```diff
@@ -57,25 +78,42 @@ public function generateUrlFunction($params, &$smarty)
             );
         }
 
+        if ($id != null && $router != null) {
+            // get url by router and id
+            /** @var Router $router */
+            $router = $this->container->get($router);
+            if (!$router) {
+                throw new \InvalidArgumentException(sprintf("Router %S doesn't exist", $router));
+            }
 
+            $excludeParams = $this->resolvePath($params, $path, $smarty);
 
-        if ($file !== null) {
-            $path = $file;
-            $mode = URL::PATH_TO_FILE;
-        } elseif ($path !== null) {
-            $mode = URL::WITH_INDEX_PAGE;
+            $url = $router->generate(
+                $id,
+                $this->getArgsFromParam($params, array_merge(['id','router'], $excludeParams))
+            );
         } else {
-            throw new \InvalidArgumentException(Translator::getInstance()->trans("Please specify either 'path' or 'file' parameter in {url} function."));
-        }
-
-        $excludeParams = $this->resolvePath($params, $path, $smarty);
+            if ($file !== null) {
+                $path = $file;
+                $mode = URL::PATH_TO_FILE;
+            } elseif ($path !== null) {
+                $mode = URL::WITH_INDEX_PAGE;
+            } else {
+                throw new \InvalidArgumentException(
+                    Translator::getInstance()->trans(
+                        "Please specify either 'path' or 'file' parameter in {url} function."
```

Comment:
Now can also specify a route id, this message should probably be updated accordingly.